File: modularmotifs/core/float.py
"""Data structure to capture the essence of a float

Possibly over-engineering here...
"""
from modularmotifs.core.motif import Color
from typing import Tuple

class FloatStrand:
    """Floats are horizontal lengths of yarn that occur because of color changes in stranded colorwork knitting
    """
    # Length of the float
    __length: int
    # the x coordinate of where the float begins. The next stitch using this color is at __start + __length + 1 -- note that this is inclusize
    __start: int
    # The row where the float occurred
    __rownum: int
    # Color
    __color: Color

    def __init__(self, start: int, length: int, row: int = -1, color: Color = None):
        """Create an instance of a FloatStrand

        Args
        ====

        start: int -- the left x coordinate of this float (i.e., the last stitch knit with this color)
        
        length: int -- the number of stitches in between that were knit in the other color. The right x coordinate of this float (i.e., the next stitch knit with this color) should be at start + length + 1

        row: int, optional -- the row where this float occurs

        color: Color, optional -- the color of this float. If provided, this color should not be 'invisible'
        """
        assert color is None or color != Color.INVIS, f"FloatStrand: color {color} should be None or not 'invisible'"
        self.__length = length
        self.__start = start
        self.__rownum = row
        self.__color = color
        pass

    def has_row(self) -> bool:
        return self.__rownum and self.__rownum != -1

    def get_row(self) -> int:
        return self.__rownum

    def has_color(self) -> bool:
        return self.__color is not None

    def get_color(self) -> Color:
        return self.__color

    def __len__(self) -> int:
        return self.__length

    def x_left(self) -> int:
        return self.__start

    def x_right(self) -> int:
        return self.__start + self.__length + 1

    def xs(self) -> Tuple[int, int]:
        return self.x_left(), self.x_right()

    
    

    
    

File: react-app/src/core/Float.ts
/**
 * Data structure to capture the essence of a float in stranded colorwork knitting.
 * Floats are horizontal lengths of yarn that occur because of color changes.
 */

import { Color } from "./Motif";

class FloatStrand {
    // Length of the float
    private length: number;
    // The x coordinate where the float begins
    private start: number;
    // The row where the float occurred
    private rownum: number;
    // Color of the float
    private color?: Color;

    /**
     * Create an instance of a FloatStrand
     * 
     * @param start - The left x coordinate of this float (i.e., the last stitch knit with this color)
     * @param length - The number of stitches in between that were knit in the other color.
     *                 The right x coordinate of this float (i.e., the next stitch knit with this color) should be at start + length + 1
     * @param row - The row where this float occurs (optional, defaults to -1)
     * @param color - The color of this float (optional, must not be Color.INVIS if provided)
     */
    constructor(start: number, length: number, row: number = -1, color?: Color) {
        if (color !== undefined && color === Color.INVIS) {
            throw new Error(`FloatStrand: color ${color} should be undefined or not 'invisible'`);
        }
        this.length = length;
        this.start = start;
        this.rownum = row;
        this.color = color;
    }

    hasRow(): boolean {
        return this.rownum !== -1;
    }

    getRow(): number {
        return this.rownum;
    }

    hasColor(): boolean {
        return this.color !== undefined;
    }

    getColor(): Color | undefined {
        return this.color;
    }

    getLength(): number {
        return this.length;
    }

    xLeft(): number {
        return this.start;
    }

    xRight(): number {
        return this.start + this.length + 1;
    }

    xs(): [number, number] {
        return [this.xLeft(), this.xRight()];
    }

    toString(): string {
        return `Float(row ${this.rownum}, ${this.start}-${this.xRight()}, ${this.length}, ${this.color})`;
    }
}
